### DISCUSSION TITLE
VbAsyncSocket-master\test\Winsock\ no event Private Sub ctxWinsock_Connect()

### DISCUSSION THREAD

**Discussion by AnzorSh** [wqweto/VbAsyncSocket#27] [Q&A] [answered]
First, thank you so much!

https://github.com/wqweto/VbAsyncSocket/blob/13532ff07311c703502deff6309e6887188da5c2/test/Winsock/Form2.frm#L91

I turned off the buttons Command1, Command2.

I'm running two Project2.exe
in the first I click HTTPS Server
in the second I execute the following lines:
```
ctxWinsock.Protocol = sckTLSProtocol
ctx Winsock.Connect "127.0.0.1", 8088 'it is mine
```

the first one has an event:
```
Private Sub ctxServer_ConnectionRequest(Index As Integer, ByVal requestID As Long)
Command2.Enabled = True 'it is mine
    Debug.Print "ctxServer_ConnectionRequest, requestID=" & requestID & ", RemoteHostIP=" & ctxServer(Index).RemoteHostIP & ", RemotePort=" & ctxServer(Index).RemotePort, Timer
```
the second one has no event:
```
Private Sub ctxWinsock_Connect()
Command1.Enabled = True 'it is mine
    Debug.Print "Connected to " & ctxWinsock.RemoteHostIP, Timer
```

If it possible, tell me how do I fix it?

**Comment by wqweto** [comment]
> If it possible, tell me how do I fix it?

What is the entire code in `ctxServer_ConnectionRequest`? You need to call `Accept` on the new `requestID` to establish connection with client.

It is when the server successfully establishes a connection with the client that you can expect to get `Connect` event raised in the client `ctxWinsock`.

Here is the entire code in the test sample
```
Private Sub ctxServer_ConnectionRequest(Index As Integer, ByVal requestID As Long)
    Debug.Print "ctxServer_ConnectionRequest, requestID=" & requestID & ", RemoteHostIP=" & ctxServer(Index).RemoteHostIP & ", RemotePort=" & ctxServer(Index).RemotePort, Timer
    Load ctxServer(ctxServer.UBound + 1)
    ctxServer(ctxServer.UBound).Protocol = ctxServer(Index).Protocol
    ctxServer(ctxServer.UBound).Accept requestID
End Sub
```

**Reply by AnzorSh** [reply]
Yes, I understand. Even the data exchange is working. ctxWinsock_Connect event does not occur.
> What is the entire code in `ctxServer_ConnectionRequest`?
```
Private Sub ctxServer_ConnectionRequest(Index As Integer, ByVal requestID As Long)
Me.Caption = "ctxServer_ConnectionRequest, requestID=" & requestID & ", RemoteHostIP=" & ctxServer(Index).RemoteHostIP & ", RemotePort=" & ctxServer(Index).RemotePort & ", Timer=" & Timer
Command2.Enabled = True
    Debug.Print "ctxServer_ConnectionRequest, requestID=" & requestID & ", RemoteHostIP=" & ctxServer(Index).RemoteHostIP & ", RemotePort=" & ctxServer(Index).RemotePort, Timer
    Load ctxServer(ctxServer.UBound + 1)
    ctxServer(ctxServer.UBound).Protocol = ctxServer(Index).Protocol
    ctxServer(ctxServer.UBound).Accept requestID
End Sub
```

**Reply by wqweto** [reply]
Try handling `Error` event too like this
```
Private Sub ctxWinsock_Error(ByVal Number As Long, Description As String, ByVal Scode As UcsErrorConstants, Source As String, HelpFile As String, ByVal HelpContext As Long, CancelDisplay As Boolean)
    Debug.Print "Error: " & Description
End Sub
```
You'll probably get `No certificate subject name matches target host name` error because you are trying to connect to loopback address `127.0.0.1` but the server is using a self-signed certificate that does not include this IP address.

I just added in commit c26bae0f3c146396dcb8684098b1d75764983ebd a parameter `LocalFeatures` to `Connect` method so that you can try connecting to local server like this
```
    ctxWinsock.Connect "127.0.0.1", 8088, ucsTlsIgnoreServerCertificateErrors
```
This will allow not checking the self-signed certificate for valid server name (and fail) so that connection is established and `Connect` event is raised.

**Comment by AnzorSh** [comment, marked as answer]
It works! Thank you so much!!!

**Comment by AnzorSh** [comment]
There was no error event.
I added
```
Private Sub ctxServer_Error(Index As Integer, ByVal Number As Long, Description As String, ByVal Scode As UcsErrorConstants, Source As String, HelpFile As String, ByVal HelpContext As Long, CancelDisplay As Boolean)
Me.Caption = "ctxServer_Error: " & Description & ", Timer=" & Timer
End Sub

Private Sub ctxWinsock_Error(ByVal Number As Long, Description As String, ByVal Scode As UcsErrorConstants, Source As String, HelpFile As String, ByVal HelpContext As Long, CancelDisplay As Boolean)
Me.Caption = "ctxWinsock_Error, Error: " & Description & ", Timer=" & Timer
    Debug.Print "Error: " & Description
End Sub
```
